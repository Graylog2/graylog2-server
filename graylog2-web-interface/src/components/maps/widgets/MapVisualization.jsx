/* eslint-env browser */
import PropTypes from 'prop-types';
import React from 'react';
import createReactClass from 'create-react-class';
import { Map, TileLayer, CircleMarker, Popup } from 'react-leaflet';
import { LatLngBounds } from 'leaflet';

import {} from 'leaflet/dist/leaflet.css';
import style from './MapVisualization.css';

const MapVisualization = createReactClass({
  displayName: 'MapVisualization',

  propTypes: {
    id: PropTypes.string.isRequired,
    data: PropTypes.object,
    height: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    url: PropTypes.string,
    attribution: PropTypes.string,
    interactive: PropTypes.bool,
    onRenderComplete: PropTypes.func,
    locked: PropTypes.bool, // Disables zoom and dragging
  },

  getDefaultProps() {
    return {
      data: {},
      url: 'https://{s}.tile.openstreetmap.org/{z}/{x}/{y}.png',
      attribution: '&copy; <a href="http://osm.org/copyright" target="_blank">OpenStreetMap</a> contributors',
      interactive: true,
      onRenderComplete: () => {},
      locked: false,
    };
  },

  getInitialState() {
    return {
      zoomLevel: 0,
    };
  },

  componentDidMount() {
    this._fitBounds();
  },

  componentDidUpdate(prevProps) {
    if (this.props.height !== prevProps.height || this.props.width !== prevProps.width) {
      this._forceMapUpdate();
      this._fitBounds();
    }
  },

  _map: undefined,
  _isMapReady: false,
  _areTilesReady: false,
  position: [0, 0],
  bounds: new LatLngBounds([-80, -170], [80, 170]),
  MARKER_RADIUS_SIZES: 10,
  MARKER_RADIUS_INCREMENT_SIZES: 10,

  // Workaround to avoid wrong placed markers or empty tiles if the map container size changed.
  _forceMapUpdate() {
    if (this._map) {
      window.dispatchEvent(new Event('resize'));
      this._map.leafletElement.invalidateSize(this.props.interactive);
    }
  },

  _fitBounds() {
    if (this._map) {
      this._map.leafletElement.fitBounds(this.bounds);
    }
  },

  // Coordinates are given as "lat,long"
  _formatMarker(coordinates, occurrences, min, max, increment) {
    const formattedCoordinates = coordinates.split(',').map(component => Number(component));
    const radius = this._getBucket(occurrences, this.MARKER_RADIUS_SIZES, min, max, increment);
    return (
      <CircleMarker key={coordinates}
                    center={formattedCoordinates}
                    radius={radius}
                    color="#AF2228"
                    fillColor="#D3242B"
                    weight={2}
                    opacity={0.8}>
        <Popup>
          <dl>
            <dt>Coordinates:</dt>
            <dd>{coordinates}</dd>
            <dt>Number of occurrences:</dt>
            <dd>{occurrences}</dd>
          </dl>
        </Popup>
      </CircleMarker>
    );
  },

  _onZoomChange(event) {
    this.setState({ zoomLevel: event.target.getZoom() });
  },

  _getBucket(value, bucketCount, minValue, maxValue, increment) {
    // Calculate bucket size based on min/max value and the number of buckets.
    const bucketSize = (maxValue - minValue) / bucketCount;
    // Select bucket for the given value.
    const bucket = value < maxValue ? Math.ceil((value - minValue) / bucketSize) : bucketCount;

    return bucket + increment;
  },

  _handleRenderComplete() {
    if (this._areTilesReady && this._isMapReady) {
      this.props.onRenderComplete();
    }
  },

  _handleMapReady() {
    this._isMapReady = true;
    this._handleRenderComplete();
  },

  _handleTilesReady() {
    this._areTilesReady = true;
    this._handleRenderComplete();
  },

  render() {
    const { data, id, height, width, url, attribution, interactive, locked } = this.props;

    const terms = data.terms;
    const occurrences = Object.keys(terms).map(k => terms[k]);
    const minOccurrences = occurrences.reduce((prev, cur) => Math.min(prev, cur), Infinity);
    const maxOccurrences = occurrences.reduce((prev, cur) => Math.max(prev, cur), -Infinity);
    const increment = this._getBucket(this.state.zoomLevel, this.MARKER_RADIUS_INCREMENT_SIZES, 1, 10, 1);

    const coordinates = Object.keys(terms);
    const markers = coordinates.map(aCoordinates => this._formatMarker(aCoordinates, terms[aCoordinates], minOccurrences, maxOccurrences, increment));

    return (
      <div className={locked ? style.mapLocked : ''}>
        {locked && <div className={style.overlay} style={{ height: height, width: width }} />}
        <Map ref={(c) => { this._map = c; }}
             id={`visualization-${id}`}
             center={this.position}
             bounds={this.bounds}
             zoom={this.state.zoomLevel}
             onZoomend={this._onZoomChange}
             className={style.map}
             style={{ height: height, width: width }}
             scrollWheelZoom={false}
             animate={interactive}
             zoomAnimation={interactive}
             fadeAnimation={interactive}
             markerZoomAnimation={interactive}
             onResize={this._fitBounds}
             whenReady={this._handleMapReady}>
          <TileLayer url={url} maxZoom={19} attribution={attribution} onLoad={this._handleTilesReady} />
          {markers}
        </Map>
      </div>
    );
  },
});

export default MapVisualization;
