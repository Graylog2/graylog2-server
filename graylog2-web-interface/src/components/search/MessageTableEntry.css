:local .timezoneInfo {
    color: #aaa;
}

:local .terminalView {
}

:local .terminalView :global(tr.message-row td) {
    padding-bottom: 0 !important;
    color: white;
}

:local .terminalView :global(.message-detail-row td) {
    background: white;
    left: 0;
}

:local .terminalView :global(.message-wrapper:after) {
    display: none;
}

:local .terminalSource {
    color: #AF892D;
}
