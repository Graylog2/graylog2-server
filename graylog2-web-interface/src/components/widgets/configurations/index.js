export { default as CountWidgetCreateConfiguration } from './CountWidgetCreateConfiguration';
export { default as CountWidgetEditConfiguration } from './CountWidgetEditConfiguration';
export { default as FieldChartWidgetConfiguration } from './FieldChartWidgetConfiguration';
export { default as QueryConfiguration } from './QueryConfiguration';
export { default as QuickValuesConfiguration } from './QuickValuesConfiguration';
export { default as QuickValuesWidgetCreateConfiguration } from './QuickValuesWidgetCreateConfiguration';
export { default as QuickValuesWidgetEditConfiguration } from './QuickValuesWidgetEditConfiguration';
export { default as QuickValuesHistogramWidgetCreateConfiguration } from './QuickValuesHistogramWidgetCreateConfiguration';
export { default as QuickValuesHistogramWidgetEditConfiguration } from './QuickValuesHistogramWidgetEditConfiguration';
export { default as StackedChartWidgetConfiguration } from './StackedChartWidgetConfiguration';
export { default as StatisticalCountWidgetCreateConfiguration } from './StatisticalCountWidgetCreateConfiguration';
export { default as StatisticalCountWidgetEditConfiguration } from './StatisticalCountWidgetEditConfiguration';
