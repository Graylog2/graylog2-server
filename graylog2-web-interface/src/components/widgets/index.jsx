import {} from './types';

export { default as Widget } from './Widget';
export { default as WidgetConfigModal } from './WidgetConfigModal';
export { default as WidgetCreationModal } from './WidgetCreationModal';
export { default as WidgetEditConfigModal } from './WidgetEditConfigModal';
export { default as WidgetFooter } from './WidgetFooter';
export { default as WidgetHeader } from './WidgetHeader';
export { default as WidgetVisualizationNotFound } from './WidgetVisualizationNotFound';
