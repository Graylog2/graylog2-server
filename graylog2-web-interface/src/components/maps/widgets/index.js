import { PluginManifest, PluginStore } from 'graylog-web-plugin/plugin';
import MapVisualization from 'components/maps/widgets/MapVisualization';

PluginStore.register(new PluginManifest({}, {
  widgets: [
    {
      type: 'org.graylog.plugins.map.widget.strategy.MapWidgetStrategy',
      displayName: 'Map',
      defaultHeight: 5,
      defaultWidth: 3,
      visualizationComponent: MapVisualization,
    },
  ],
}));

