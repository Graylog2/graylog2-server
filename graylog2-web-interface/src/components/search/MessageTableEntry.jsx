import PropTypes from 'prop-types';
import React from 'react';
import Immutable from 'immutable';
import ImmutablePropTypes from 'react-immutable-proptypes';
import { Popover, OverlayTrigger } from 'react-bootstrap';

import { Timestamp } from 'components/common';
import StringUtils from 'util/StringUtils';
import DateTime from 'logic/datetimes/DateTime';
import DecorationStats from 'logic/message/DecorationStats';
import MessageDetail from './MessageDetail';
import style from './MessageTableEntry.css';

class MessageTableEntry extends React.Component {
  static propTypes = {
    allStreams: ImmutablePropTypes.list.isRequired,
    allStreamsLoaded: PropTypes.bool.isRequired,
    disableSurroundingSearch: PropTypes.bool,
    expandAllRenderAsync: PropTypes.bool.isRequired,
    expanded: PropTypes.bool.isRequired,
    highlight: PropTypes.bool,
    highlightMessage: PropTypes.string,
    inputs: ImmutablePropTypes.map.isRequired,
    message: PropTypes.shape({
      fields: PropTypes.object.isRequired,
      highlight_ranges: PropTypes.object,
      id: PropTypes.string.isRequired,
      index: PropTypes.string.isRequired,
    }).isRequired,
    nodes: ImmutablePropTypes.map.isRequired,
    searchConfig: PropTypes.object,
    selectedFields: ImmutablePropTypes.orderedSet,
    showMessageRow: PropTypes.bool,
    streams: ImmutablePropTypes.map.isRequired,
    toggleDetail: PropTypes.func.isRequired,
  };

  static defaultProps = {
    disableSurroundingSearch: false,
    highlight: false,
    highlightMessage: undefined,
    searchConfig: undefined,
    selectedFields: Immutable.OrderedSet(),
    showMessageRow: false,
  };

  shouldComponentUpdate(newProps) {
    if (this.props.highlight !== newProps.highlight) {
      return true;
    }
    if (!Immutable.is(this.props.selectedFields, newProps.selectedFields)) {
      return true;
    }
    if (this.props.expanded !== newProps.expanded) {
      return true;
    }
    if (this.props.expandAllRenderAsync !== newProps.expandAllRenderAsync) {
      return true;
    }
    if (this.props.allStreamsLoaded !== newProps.allStreamsLoaded) {
      return true;
    }
    if (this.props.showMessageRow !== newProps.showMessageRow) {
      return true;
    }
    return false;
  }

  possiblyHighlight = (fieldName, fullOrigValue, truncate) => {
    // Ensure the field is a string for later processing
    const fullStringOrigValue = StringUtils.stringify(fullOrigValue);

    // Truncate the field to 2048 characters if requested. This is for performance reasons to avoid hogging the CPU.
    // It's not optimal, more like a workaround to at least being able to show the page...
    const origValue = truncate ? fullStringOrigValue.slice(0, 2048) : fullStringOrigValue;

    if (this.props.highlight && this.props.message.highlight_ranges) {
      if (this.props.message.highlight_ranges.hasOwnProperty(fieldName)) {
        const chunks = [];
        const highlights = Immutable.fromJS(this.props.message.highlight_ranges[fieldName]).sortBy(range => range.get('start'));
        let position = 0;
        let key = 0;
        highlights.forEach((range, idx) => {
          if (position !== range.get('start')) {
            chunks.push(<span key={key++}>{origValue.substring(position, range.get('start'))}</span>);
          }
          chunks.push(<span key={key++} className="result-highlight-colored">{origValue.substring(range.get('start'), range.get('start') + range.get('length'))}</span>);
          if ((idx + 1) < highlights.size) {
            const nextRange = highlights.get(idx + 1);
            chunks.push(<span key={key++}>{origValue.substring(range.get('start') + range.get('length'), nextRange.get('start'))}</span>);
            position = nextRange.get('start');
          } else {
            chunks.push(<span key={key++}>{origValue.substring(range.get('start') + range.get('length'))}</span>);
            position = range.get('start') + range.get('length');
          }
        });
        return <span>{chunks}</span>;
      }
      return String(origValue);
    }
    return String(origValue);
  };

  _toggleDetail = () => {
    this.props.toggleDetail(`${this.props.message.index}-${this.props.message.id}`);
  };

  _toTimestamp = (value) => {
    const popoverHoverFocus = (
      <Popover id="popover-trigger-hover-focus">
        This timestamp is rendered in your timezone.
      </Popover>
    );

    return (
      <span>
        <Timestamp dateTime={value} format={DateTime.Formats.TIMESTAMP_TZ} />
        <OverlayTrigger trigger={['hover']} overlay={popoverHoverFocus} >
          <i className={`fa fa-fw fa-info ${style.timezoneInfo}`} />
        </OverlayTrigger>
      </span>
    );
  };

  renderForDisplay = (fieldName, truncate) => {
    const fullOrigValue = this.props.message.fields[fieldName];
    const isDecorated = DecorationStats.isFieldDecorated(this.props.message, fieldName);

    if (isDecorated && (typeof fullOrigValue === 'object') && (fullOrigValue.type === 'a')) {
      const link = fullOrigValue.href;
      return React.createElement('a', { href: link }, link);
    }

    if (fullOrigValue === undefined) {
      return '';
    }

    /* Timestamp can not be highlighted by elastic search. So we can safely
     * skip them from highlighting. */
    if (fieldName === 'timestamp') {
      return this._toTimestamp(fullOrigValue);
    } else {
      return this.possiblyHighlight(fieldName, fullOrigValue, truncate);
    }
  };

  render() {
    const colSpanFixup = this.props.selectedFields.size + 1;

    let classes = 'message-group';
    if (this.props.expanded) {
      classes += ' message-group-toggled';
    }
    if (this.props.message.id === this.props.highlightMessage) {
      classes += ' message-highlight';
    }

    return (
      <tbody className={classes}>
        <tr className="fields-row" onClick={this._toggleDetail}>
          <td><strong>
            <Timestamp dateTime={this.props.message.fields.timestamp} />
          </strong></td>
          { this.props.selectedFields.toSeq().map(selectedFieldName => (<td
            key={selectedFieldName}>{this.renderForDisplay(selectedFieldName, true)} </td>)) }
        </tr>

        {this.props.showMessageRow &&
        <tr className="message-row" onClick={this._toggleDetail}>
          <td colSpan={colSpanFixup}><div className="message-wrapper">{this.renderForDisplay('message', true)}</div></td>
        </tr>
        }
        {this.props.expanded &&
        <tr className="message-detail-row" style={{ display: 'table-row' }}>
          <td colSpan={colSpanFixup}>
            <MessageDetail message={this.props.message}
                           inputs={this.props.inputs}
                           streams={this.props.streams}
                           allStreams={this.props.allStreams}
                           allStreamsLoaded={this.props.allStreamsLoaded}
                           nodes={this.props.nodes}
                           renderForDisplay={this.renderForDisplay}
                           disableSurroundingSearch={this.props.disableSurroundingSearch}
                           expandAllRenderAsync={this.props.expandAllRenderAsync}
                           searchConfig={this.props.searchConfig} />
          </td>
        </tr>
        }
      </tbody>
    );
  }
}

export default MessageTableEntry;
