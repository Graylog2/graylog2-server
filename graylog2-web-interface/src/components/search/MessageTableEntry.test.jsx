import React from 'react';
import renderer from 'react-test-renderer';
import Immutable from 'immutable';
import DateTime from 'logic/datetimes/DateTime';
import {} from 'jquery-ui/ui/version';
import {} from 'jquery-ui/ui/effect';
import {} from 'jquery-ui/ui/plugin';
import {} from 'jquery-ui/ui/widget';
import {} from 'jquery-ui/ui/widgets/mouse';
import { mount } from 'enzyme';

import MessageTableEntry from 'components/search/MessageTableEntry';

describe('<MessageTableEntry />', () => {
  DateTime.getBrowserTimezone = () => { return 'Europe/Berlin'; };
  const allStreams = Immutable.List([
    { id: '01', description: 'stream1' },
    { id: '02', description: 'stream2' },
  ]);
  const streams = Immutable.Map({ '01': { id: '01', description: 'stream1' } });
  const inputs = Immutable.Map({ '00001': { id: '00001', title: 'syslog', name: 'syslog' } });
  const message = {
    fields: { message: '2018-01-22T15:36:02.189Z foo', timestamp: '2018-01-22T15:36:02.189Z' },
    formatted_fields: { message: '2018-01-22T15:36:02.189Z foo', timestamp: '2018-01-22T15:36:02.189Z' },
    highlight_ranges: {},
    id: '01',
    index: '01',
  };
  const nodes = Immutable.Map({});

  describe('rendering', () => {
    it('should render a MessageTableEntry', () => {
      const wrapper = renderer.create(<MessageTableEntry
        allStreams={allStreams}
        allStreamsLoaded
        disableSurroundingSearch
        expandAllRenderAsync={false}
        expanded
        inputs={inputs}
        searchConfig={{}}
        message={message}
        nodes={nodes}
        streams={streams}
        toggleDetail={jest.fn}
      />);
      expect(wrapper.toJSON()).toMatchSnapshot();
    });
  });

  describe('timezone handling', () => {
    it('should render a in the UTC timezone', () => {
      const wrapper = mount(<table><MessageTableEntry
        allStreams={allStreams}
        allStreamsLoaded
        disableSurroundingSearch
        expandAllRenderAsync={false}
        expanded
        inputs={inputs}
        searchConfig={{}}
        message={message}
        nodes={nodes}
        streams={streams}
        toggleDetail={jest.fn}
      /></table>);
      expect(wrapper.find('time').at(1).text()).toEqual('2018-01-22 16:36:02.189 +01:00');
    });

    it('should render a in the USA/Honolulu timezone', () => {
      DateTime.getBrowserTimezone = () => { return 'Pacific/Honolulu'; };
      const wrapper = mount(<table><MessageTableEntry
        allStreams={allStreams}
        allStreamsLoaded
        disableSurroundingSearch
        expandAllRenderAsync={false}
        expanded
        inputs={inputs}
        searchConfig={{}}
        message={message}
        nodes={nodes}
        streams={streams}
        toggleDetail={jest.fn}
      /></table>);
      expect(wrapper.find('time').at(1).text()).toEqual('2018-01-22 05:36:02.189 -10:00');
    });
  });
});
