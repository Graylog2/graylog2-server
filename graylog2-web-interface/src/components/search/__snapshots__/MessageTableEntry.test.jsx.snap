// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MessageTableEntry /> rendering should render a MessageTableEntry 1`] = `
<tbody
  className="message-group message-group-toggled"
>
  <tr
    className="fields-row"
    onClick={[Function]}
  >
    <td>
      <strong>
        <time
          dateTime="2018-01-22T15:36:02.189Z"
          title="2018-01-22T15:36:02.189Z"
        >
          2018-01-22 16:36:02.189
        </time>
      </strong>
    </td>
  </tr>
  <tr
    className="message-detail-row"
    style={
      Object {
        "display": "table-row",
      }
    }
  >
    <td
      colSpan={1}
    >
      <div>
        <div
          className="row-sm row"
        >
          <div
            className="col-md-12"
          >
            <div
              className="pull-right btn-group btn-group-sm"
            >
              <a
                className="btn btn-default"
                href="/messages/01/01"
                onClick={[Function]}
                onKeyDown={[Function]}
              >
                Permalink
              </a>
              <button
                className="btn btn-default"
                data-clipboard-action="copy"
                data-clipboard-button={true}
                data-clipboard-text="01"
                disabled={false}
                onClick={[Function]}
                style={undefined}
                title={undefined}
                type="button"
              >
                Copy ID
              </button>
              <div
                className="dropdown btn-group btn-group-sm"
              >
                <button
                  aria-expanded={false}
                  aria-haspopup={true}
                  className="dropdown-toggle btn btn-sm btn-default"
                  disabled={false}
                  id="select-stream-dropdown"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  type="button"
                >
                  Test against stream
                   
                  <span
                    className="caret"
                  />
                </button>
                <ul
                  aria-labelledby="select-stream-dropdown"
                  className="dropdown-menu dropdown-menu-right"
                  role="menu"
                >
                  <li
                    className=""
                    role="presentation"
                    style={undefined}
                  >
                    <a
                      action="push"
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    />
                  </li>
                  <li
                    className=""
                    role="presentation"
                    style={undefined}
                  >
                    <a
                      action="push"
                      href="#"
                      onClick={[Function]}
                      onKeyDown={[Function]}
                      role="menuitem"
                      tabIndex="-1"
                    />
                  </li>
                </ul>
              </div>
            </div>
            <h3
              className="message-details-title"
            >
              <i
                className="fa fa-envelope"
              />
              Â 
              <a
                onClick={[Function]}
                style={Object {}}
              >
                01
              </a>
            </h3>
          </div>
        </div>
        <div
          className="row"
        >
          <div
            className="col-md-3"
          >
            <dl
              className="message-details"
            >
              <dt>
                Stored in index
              </dt>
              <dd>
                01
              </dd>
            </dl>
          </div>
          <div
            className="col-md-9"
          >
            <div>
              <dl
                className="message-details message-details-fields"
              >
                <span>
                  <dt>
                    message
                  </dt>
                  <dd
                    className="message-field"
                  >
                    <div
                      className="message-field-actions pull-right"
                    >
                      <div
                        className="dropdown btn-group btn-group-xs"
                      >
                        <button
                          className="btn btn-xs btn-default"
                          disabled={false}
                          onClick={[Function]}
                          type="button"
                        >
                          <i
                            className="fa fa-search-plus"
                          />
                        </button>
                        <button
                          aria-expanded={false}
                          aria-haspopup={true}
                          aria-label={
                            <i
                              className="fa fa-search-plus"
                            />
                          }
                          className="dropdown-toggle btn btn-xs btn-default"
                          disabled={false}
                          id="more-actions-dropdown-field-message"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          role="button"
                          type="button"
                        >
                           
                          <span
                            className="caret"
                          />
                        </button>
                        <ul
                          aria-labelledby="more-actions-dropdown-field-message"
                          className="dropdown-menu dropdown-menu-right"
                          role="menu"
                        >
                          <li
                            className="dropdown-submenu left-submenu"
                            onKeyDown={[Function]}
                            onSelect={[Function]}
                          >
                            <a
                              href="#"
                            >
                              Create extractor for field 
                              message
                            </a>
                            <ul
                              className="dropdown-menu"
                            >
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=copy_input&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Copy input
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=grok&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Grok pattern
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=json&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  JSON
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=regex&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Regular expression
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=regex_replace&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Replace with regular expression
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=split_and_index&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Split & Index
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=substring&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Substring
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=lookup_table&field=message&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Lookup Table
                                </a>
                              </li>
                            </ul>
                          </li>
                          <li
                            className=""
                            role="presentation"
                            style={undefined}
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              Show terms of 
                              <em>
                                message
                              </em>
                            </a>
                          </li>
                        </ul>
                      </div>
                    </div>
                    <div
                      className="field-value"
                    >
                      2018-01-22T15:36:02.189Z foo
                    </div>
                  </dd>
                </span>
                <span>
                  <dt>
                    timestamp
                  </dt>
                  <dd
                    className=""
                  >
                    <div
                      className="message-field-actions pull-right"
                    >
                      <div
                        className="dropdown btn-group btn-group-xs"
                      >
                        <button
                          className="btn btn-xs btn-default"
                          disabled={false}
                          onClick={[Function]}
                          type="button"
                        >
                          <i
                            className="fa fa-search-plus"
                          />
                        </button>
                        <button
                          aria-expanded={false}
                          aria-haspopup={true}
                          aria-label={
                            <i
                              className="fa fa-search-plus"
                            />
                          }
                          className="dropdown-toggle btn btn-xs btn-default"
                          disabled={false}
                          id="more-actions-dropdown-field-timestamp"
                          onClick={[Function]}
                          onKeyDown={[Function]}
                          role="button"
                          type="button"
                        >
                           
                          <span
                            className="caret"
                          />
                        </button>
                        <ul
                          aria-labelledby="more-actions-dropdown-field-timestamp"
                          className="dropdown-menu dropdown-menu-right"
                          role="menu"
                        >
                          <li
                            className="dropdown-submenu left-submenu"
                            onKeyDown={[Function]}
                            onSelect={[Function]}
                          >
                            <a
                              href="#"
                            >
                              Create extractor for field 
                              timestamp
                            </a>
                            <ul
                              className="dropdown-menu"
                            >
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=copy_input&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Copy input
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=grok&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Grok pattern
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=json&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  JSON
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=regex&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Regular expression
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=regex_replace&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Replace with regular expression
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=split_and_index&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Split & Index
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=substring&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Substring
                                </a>
                              </li>
                              <li
                                className=""
                                role="presentation"
                                style={undefined}
                              >
                                <a
                                  href="/system/inputs/undefined/undefined/extractors/new?extractor_type=lookup_table&field=timestamp&example_index=01&example_id=01"
                                  onClick={[Function]}
                                  onKeyDown={[Function]}
                                  role="menuitem"
                                  tabIndex="-1"
                                >
                                  Lookup Table
                                </a>
                              </li>
                            </ul>
                          </li>
                          <li
                            className=""
                            role="presentation"
                            style={undefined}
                          >
                            <a
                              href="#"
                              onClick={[Function]}
                              onKeyDown={[Function]}
                              role="menuitem"
                              tabIndex="-1"
                            >
                              Show terms of 
                              <em>
                                timestamp
                              </em>
                            </a>
                          </li>
                        </ul>
                      </div>
                    </div>
                    <div
                      className="field-value"
                    >
                      <span>
                        <time
                          dateTime="2018-01-22T15:36:02.189Z"
                          title="2018-01-22T15:36:02.189Z"
                        >
                          2018-01-22 16:36:02.189 +01:00
                        </time>
                        <i
                          className="fa fa-fw fa-info timezoneInfo"
                          onClick={null}
                          onMouseOut={[Function]}
                          onMouseOver={[Function]}
                        />
                      </span>
                    </div>
                  </dd>
                </span>
              </dl>
            </div>
          </div>
        </div>
      </div>
    </td>
  </tr>
</tbody>
`;
